transport/x/cherami/inbound.go

```diff
@@ -0,0 +1,199 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cherami
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/errors"
+	"go.uber.org/yarpc/internal/sync"
+	"go.uber.org/yarpc/serialize"
+	"go.uber.org/yarpc/transport/x/cherami/internal"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/uber/cherami-client-go/client/cherami"
+)
+
+const (
+	transportName = "cherami"
+
+	defaultPrefetchCount = 10
+
+	defaultCheramiTimeout = 15 * time.Second
+)
+
+// InboundConfig defines the config in order to create a Inbound.
+//
+// PrefetchCount controls the number of messages to buffer locally. Inbounds
+// which process messages very fast may want to specify larger value for
+// PrefetchCount for faster throughput. On the flip side larger values for
+// PrefetchCount will result in more messages being buffered locally causing
+// high memory footprint.
+type InboundConfig struct {
+	Destination   string
+	ConsumerGroup string
+	PrefetchCount int
+	Timeout       time.Duration
+}
+
+// Inbound receives Oneway YARPC requests over Cherami.
+type Inbound struct {
+	config        InboundConfig
+	consumer      cherami.Consumer
+	router        transport.Router
+	tracer        opentracing.Tracer
+	client        cherami.Client
+	clientFactory internal.ClientFactory
+
+	once sync.LifecycleOnce
+}
+
+// NewInbound builds a new Cherami inbound.
+func (t *Transport) NewInbound(config InboundConfig) *Inbound {
+	if config.PrefetchCount == 0 {
+		config.PrefetchCount = defaultPrefetchCount
+	}
+	if config.Timeout/time.Second <= 0 {
+		config.Timeout = defaultCheramiTimeout
+	}
+	return &Inbound{
+		config:        config,
+		tracer:        t.tracer,
+		client:        t.client,
+		clientFactory: t.clientFactory,
+	}
+}
+
+// Transports returns nil for now
+func (i *Inbound) Transports() []transport.Transport {
+	return nil
+}
+
+// SetRouter configures a router to handle incoming requests.
+// This satisfies the transport.Inbound interface, and would be called
+// by a dispatcher when it starts.
+func (i *Inbound) SetRouter(router transport.Router) {
+	i.router = router
+}
+
+// IsRunning returns whether the inbound is still processing requests.
+func (i *Inbound) IsRunning() bool {
+	return i.once.IsRunning()
+}
+
+// Start starts the inbound, reading and handling messages from cherami
+func (i *Inbound) Start() error {
+	return i.once.Start(i.start)
+}
+
+func (i *Inbound) start() error {
+	if i.router == nil {
+		return errors.ErrNoRouter
+	}
+
+	consumer, ch, err := i.clientFactory.GetConsumer(i.client, internal.ConsumerConfig{
+		Destination:   i.config.Destination,
+		ConsumerGroup: i.config.ConsumerGroup,
+		PrefetchCount: i.config.PrefetchCount,
+		Timeout:       i.config.Timeout,
+	})
+	if err != nil {
+		return err
+	}
+
+	i.consumer = consumer
+
+	go func() {
+		for delivery := range ch {
```

Comment:
I don't feel strongly about this but this isn't necessarily a small or
uncomplicated function. It's looping over a channel and it has multiple error
cases being handled. Moving it to its own function will make it more future-
proof. Another option is to move just the logic inside the loop to its own
function:

```
go func() {
    for delivery := range ch {
        i.handleDelivery(delivery)
    }
}()
```

I'm especially wary of functions like this when there is a chance that they
will accidentally capture and use more variables from the surrounding closure
than was originally intended. Having a separate function makes it especially
clear exactly which variables from the surrounding scope are meant for that
logic.